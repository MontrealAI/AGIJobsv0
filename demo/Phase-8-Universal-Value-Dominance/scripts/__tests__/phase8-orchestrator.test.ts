require("ts-node/register/transpile-only");

const { describe, beforeAll, expect, it } = require("@jest/globals");
const { mkdtempSync, readFileSync, rmSync } = require("node:fs");
const { tmpdir } = require("node:os");
const { join } = require("node:path");

const {
  buildSafeTransactions,
  calldata,
  computeMetrics,
  flattenCalldataEntries,
  guardrailDiagnostics,
  loadConfig,
  parseManifest,
  resolveEnvironment,
  schedulePlaybooks,
  telemetryMarkdown,
  mermaid,
  writeArtifacts,
} = require("../run-phase8-demo");

describe("Phase 8 orchestration console", () => {
  let config;

  beforeAll(() => {
    config = loadConfig();
  });

  it("parses the manifest with strict validation", () => {
    expect(config.domains).toHaveLength(5);
    expect(config.sentinels).toHaveLength(3);
    expect(config.capitalStreams).toHaveLength(3);

    const broken = JSON.parse(JSON.stringify(config));
    broken.domains[0].slug = "";
    expect(() => parseManifest(broken)).toThrowError(/domains\.0\.slug: Domain slug is required/);
  });

  it("enforces guardian coverage and autonomy guardrails in the manifest", () => {
    const guardrailBreach = JSON.parse(JSON.stringify(config));
    guardrailBreach.selfImprovement.autonomyGuards.maxAutonomyBps = 1000;
    expect(() => parseManifest(guardrailBreach)).toThrowError(
      /domains\.0\.autonomyLevelBps: Domain autonomy 7800bps exceeds guardrail cap 1000bps/, 
    );

    const missingGuard = JSON.parse(JSON.stringify(config));
    delete missingGuard.selfImprovement.autonomyGuards;
    expect(() => parseManifest(missingGuard)).toThrowError(
      /selfImprovement\.autonomyGuards\.maxAutonomyBps: Autonomy guard maxAutonomyBps is required/, 
    );

    const coverageBreach = JSON.parse(JSON.stringify(config));
    coverageBreach.sentinels = coverageBreach.sentinels.map((entry) => ({
      ...entry,
      coverageSeconds: 10,
    }));
    expect(() => parseManifest(coverageBreach)).toThrowError(
      /sentinels: Total sentinel coverage 30s is below guardian review window 720s/, 
    );
  });

  it("encodes calldata and flattens manifests deterministically", () => {
    const calls = calldata(config);
    expect(calls.setGlobalParameters).toMatch(/^0x[0-9a-f]+$/i);
    expect(calls.registerDomains).toHaveLength(config.domains.length);

    const flattened = flattenCalldataEntries(calls);
    const registerDomain = flattened.find((entry) => entry.label === "registerDomain");
    expect(registerDomain).toBeDefined();
    expect(registerDomain?.data).toMatch(/^0x[0-9a-f]+$/i);
    expect(flattened.length).toBeGreaterThan(config.domains.length + config.sentinels.length + config.capitalStreams.length);

    const safeTransactions = buildSafeTransactions(flattened, "0x1234567890abcdef1234567890abcdef12345678");
    expect(safeTransactions[0]).toMatchObject({ to: "0x1234567890abcdef1234567890abcdef12345678", value: "0" });
    expect(new Set(safeTransactions.map((tx) => tx.data)).size).toBeGreaterThan(0);
  });

  it("resolves environment overrides with validation", () => {
    expect(resolveEnvironment({})).toEqual({
      chainId: 1,
      managerAddress: "0x0000000000000000000000000000000000000000",
    });

    expect(
      resolveEnvironment({
        PHASE8_CHAIN_ID: "777",
        PHASE8_MANAGER_ADDRESS: "0xABCDEFabcdefABCDEFabcdefABCDEFabcdefabcd",
      }),
    ).toEqual({
      chainId: 777,
      managerAddress: "0xabcdefabcdefabcdefabcdefabcdefabcdefabcd",
    });

    expect(() =>
      resolveEnvironment({
        PHASE8_CHAIN_ID: "-5",
        PHASE8_MANAGER_ADDRESS: "not-an-address",
      }),
    ).toThrowError(/Environment validation failed:[\s\S]*chainId: Chain ID must be positive/);
  });

  it("writes artifact kit with telemetry and mermaid snapshots", () => {
    const metrics = computeMetrics(config);
    const data = calldata(config);
    const env = { chainId: 8453, managerAddress: "0x9999999999999999999999999999999999999999" };
    const tempDir = mkdtempSync(join(tmpdir(), "phase8-test-"));

    try {
      const outputs = writeArtifacts(config, metrics, data, env, { outputDir: tempDir });
      expect(outputs).toHaveLength(4);

      const telemetryPath = join(tempDir, "phase8-telemetry-report.md");
      const mermaidPath = join(tempDir, "phase8-mermaid-diagram.mmd");
      const telemetry = readFileSync(telemetryPath, "utf-8");
      const diagram = readFileSync(mermaidPath, "utf-8");

      expect(telemetryMarkdown(config, metrics)).toContain("Phase 8 â€” Universal Value Dominance Telemetry");
      const stableTelemetry = telemetry.replace(/Generated: .*/u, "Generated: <timestamp>");
      expect(stableTelemetry).toMatchSnapshot();
      expect(mermaid(config)).toEqual(diagram);
      expect(diagram).toMatchSnapshot();

      const safeBatch = JSON.parse(readFileSync(join(tempDir, "phase8-safe-transaction-batch.json"), "utf-8"));
      expect(safeBatch.chainId).toBe(String(env.chainId));
      expect(safeBatch.meta.createdFromSafeAddress).toBe(env.managerAddress);
    } finally {
      rmSync(tempDir, { recursive: true, force: true });
    }
  });

  it("schedules playbooks with guardrails and handles manual workflows", () => {
    const schedules = schedulePlaybooks(config, 1_700_000_000);
    expect(schedules).toHaveLength(2);
    expect(schedules[0]).toMatchObject({
      name: "Hyperparameter Evolution",
      intervalSeconds: 7 * 24 * 60 * 60,
      requiresManualScheduling: false,
    });
    expect(Date.parse(schedules[1]?.nextRun ?? "")).toBeGreaterThan(0);

    const manualConfig = JSON.parse(JSON.stringify(config));
    manualConfig.selfImprovement.playbooks[0].automation = "manual";
    const manualSchedules = schedulePlaybooks(manualConfig, 1_700_000_000);
    expect(manualSchedules[0]).toMatchObject({ requiresManualScheduling: true });
  });

  it("produces operator diagnostics for domains missing guardrails", () => {
    const stressed = JSON.parse(JSON.stringify(config));
    stressed.domains[1].resilienceIndex = 0.4;
    stressed.domains[1].heartbeatSeconds = stressed.global.heartbeatSeconds + 120;
    stressed.selfImprovement.autonomyGuards.maxAutonomyBps = stressed.domains[1].autonomyLevelBps - 500;
    stressed.sentinels = stressed.sentinels.map((entry) => ({
      ...entry,
      coverageSeconds: entry.domains.includes(stressed.domains[1].slug) ? 60 : entry.coverageSeconds,
    }));

    const diagnostics = guardrailDiagnostics(stressed);
    expect(diagnostics.some((line) => line.includes("coverage"))).toBe(true);
    expect(diagnostics.some((line) => line.includes("resilience"))).toBe(true);
    expect(diagnostics.some((line) => line.includes("heartbeat"))).toBe(true);
    expect(diagnostics.some((line) => line.includes("autonomy"))).toBe(true);
  });
});
