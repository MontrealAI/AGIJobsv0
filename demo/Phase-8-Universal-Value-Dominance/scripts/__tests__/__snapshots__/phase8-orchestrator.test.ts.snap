// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Phase 8 orchestration console writes artifact kit with telemetry and mermaid snapshots 1`] = `
"# Phase 8 — Universal Value Dominance Telemetry
Generated: <timestamp>

## Global Metrics
- Total monthly value flow: $688.00B
- Annual capital allocation: $1.97T
- Average resilience index: 0.925
- Sentinel coverage per guardian cycle: 18.0 minutes
- Domains covered by sentinels: 100.0%
- Maximum encoded autonomy: 7800 bps
- Self-improvement cadence: every 2.00 hours
- Last self-improvement execution: 2023-11-14T22:20:00.000Z

## Diagnostics
- Planetary Finance Mesh has 360s of sentinel coverage but guardian review window requires 720s
- Climate Harmonizer Array has 420s of sentinel coverage but guardian review window requires 720s
- Health Sovereign Continuum has 300s of sentinel coverage but guardian review window requires 720s
- Infrastructure Synthesis Grid has 420s of sentinel coverage but guardian review window requires 720s
- Knowledge Lattice Nexus has 360s of sentinel coverage but guardian review window requires 720s

## Governance Control Surface
- Treasury: 0x8a1f8f5e3b9d9b2a7b6c5e4f3d2c1b0a9f8e7d6c
- Universal vault: 0xb5c7fe65f0c84a22330a3e913450b1214cc0a8f9
- Upgrade coordinator: 0x3e8b71da4c5e981a2d4facfe97b53bd2736d1d10
- Validator registry: 0x21f49cff41f9e69a13a3d6cef7b1a92d9c4e0d8e
- Mission control: 0xe9a6b9b6c6a53382e5e7d1b8e0f93c5f4a32e1a4
- Knowledge graph: 0x9b2b5f9c4d4e58a63e934f52aa4c3f62f6c4b672
- Guardian council: 0x4c3ab8173d97d58b0daa9f73a2e3e87a4fe98c87
- System pause: 0xdd1f26bcb5d0faa1a69729db849bb4c276c74e5c
- Manifest URI: ipfs://agi-jobs/phase8/manifest.json
- Max drawdown guard: 3200 bps

## Domains
| Domain | Autonomy (bps) | Resilience | Heartbeat (s) | TVL cap | Monthly value | Sentinels | Capital streams |
| --- | --- | --- | --- | --- | --- | --- | --- |
| Planetary Finance Mesh | 7800 | 0.960 | 300 | 420… (10^24) | $245.00B | Capital Watch Exocomptroller | Planetary Resilience Fund |
| Climate Harmonizer Array | 7400 | 0.940 | 180 | 185… (10^24) | $128.00B | Solar Shield Guardian | Climate Stabilization Endowment |
| Health Sovereign Continuum | 7000 | 0.910 | 240 | 920… (10^23) | $94.00B | Bio Sentinel Continuity | Planetary Resilience Fund |
| Infrastructure Synthesis Grid | 7600 | 0.915 | 210 | 215… (10^24) | $167.00B | Solar Shield Guardian | Climate Stabilization Endowment, Innovation Thrust Catalyst |
| Knowledge Lattice Nexus | 6900 | 0.902 | 150 | 680… (10^23) | $54.00B | Capital Watch Exocomptroller | Planetary Resilience Fund, Innovation Thrust Catalyst |

## Sentinel Lattice
- Solar Shield Guardian · coverage 420s · sensitivity 450 bps · guarding climate-harmonizer, infrastructure-synthesis
- Capital Watch Exocomptroller · coverage 360s · sensitivity 380 bps · guarding planetary-finance, knowledge-lattice
- Bio Sentinel Continuity · coverage 300s · sensitivity 420 bps · guarding health-sovereign

## Capital Streams
- Climate Stabilization Endowment · $720.00B/yr · expansion 1300 bps · targets climate-harmonizer, infrastructure-synthesis
- Planetary Resilience Fund · $890.00B/yr · expansion 1250 bps · targets planetary-finance, health-sovereign, knowledge-lattice
- Innovation Thrust Catalyst · $360.00B/yr · expansion 1600 bps · targets knowledge-lattice, infrastructure-synthesis

## Self-Improvement Kernel
- Strategic plan: cadence 7200s (2.00 h) · hash 0x7d2556f5ae0a7dd2e9c3c5e29377b7af168e2b2f42a3e08d96b9f8e32d4b9f10 · last report ipfs://agi-jobs/phase8/self-improvement/reports/epoch-144.json
- Playbook Hyperparameter Evolution (weekly) · owner 0x4c3ab8173d97d58b0daa9f73a2e3e87a4fe98c87 · guardrails checksum-verification, zk-proof-of-alignment · Next run: 2023-11-21T22:20:00.000Z
- Playbook Universal Value Mesh Stress Tests (hourly) · owner 0x3e8b71da4c5e981a2d4facfe97b53bd2736d1d10 · guardrails branch-protected-ci, multi-agent-cross-check · Next run: 2023-11-14T23:20:00.000Z
- Autonomy guard: ≤8000 bps · override window 15 minutes · escalation guardian-council → dao-emergency → sentinel-lockdown"
`;

exports[`Phase 8 orchestration console writes artifact kit with telemetry and mermaid snapshots 2`] = `
"graph TD
  Governance[[Guardian DAO]] --> Manager(Phase8UniversalValueManager)
  Manager --> planetaryfinance([Planetary Finance Mesh])
  Manager --> climateharmonizer([Climate Harmonizer Array])
  Manager --> healthsovereign([Health Sovereign Continuum])
  Manager --> infrastructuresynthesis([Infrastructure Synthesis Grid])
  Manager --> knowledgelattice([Knowledge Lattice Nexus])
  Manager --> Treasury[[Universal Treasury]]
  Manager --> Sentinels{Sentinel lattice}
  Sentinels --> Pause[System pause]
  Manager --> Streams[[Capital Streams]]
  Streams --> climatestabilizationStream([Climate Stabilization Endowment])
  Streams --> planetaryresilienceStream([Planetary Resilience Fund])
  Streams --> innovationthrustStream([Innovation Thrust Catalyst])
  Streams --> Domains([Autonomous domains])"
`;
